refactor: Make m_cs_fee_estimator non-recursive

hebasto opened the pull request:
This PR eliminates the only place that `m_cs_fee_estimator` is recursively locked by refactoring out `_removeTx` member function.

Related to #19303.

hebasto committed:
Add thread safety annotations to CBlockPolicyEstimator public functions

hebasto committed:
refactor: Add non-thread-safe CBlockPolicyEstimator::_removeTx helper

This changes removes recursion in the m_cs_fee_estimator locks.

hebasto committed:
refactor: Make m_cs_fee_estimator non-recursive

MarcoFalke commented on src/policy/fees.h:
```diff
@@ -267,6 +279,10 @@ class CBlockPolicyEstimator
     unsigned int MaxUsableEstimate() const EXCLUSIVE_LOCKS_REQUIRED(m_cs_fee_estimator);
+
+    /** A non-thread-safe helper for the removeTx function */
+    bool removeTxNonThreadSafeHelper(const uint256& hash, bool inBlock)
```
```suggestion
    bool _removeTx(const uint256& hash, bool inBlock)
```

According to similar methods:

```cpp
    void _RelayTransaction(const uint256& txid, const uint256& wtxid)
        EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```

hebasto replied:
Thanks! It's great to have a common approach for such a naming case. [Updated](https://github.com/bitcoin/bitcoin/pull/22014#issuecomment-845784457).

maflcko reviewed (commented):
haven't reviewed, but I left a style comment

hebasto commented:
Updated 5efec949265fd2a04f131a2a454ec984b033eb66 -> 8c277b19c8f262e550cffe263e6d910b687ac882 ([pr22014.01](https://github.com/hebasto/bitcoin/commits/pr22014.01) -> [pr22014.02](https://github.com/hebasto/bitcoin/commits/pr22014.02), [diff](https://github.com/hebasto/bitcoin/compare/pr22014.01..pr22014.02)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/22014#discussion_r636735149)

hebasto commented on src/policy/fees.cpp:
```diff
@@ -556,7 +562,8 @@ void CBlockPolicyEstimator::processTransaction(const CTxMemPoolEntry& entry, boo
 bool CBlockPolicyEstimator::processBlockTx(unsigned int nBlockHeight, const CTxMemPoolEntry* entry)
 {
-    if (!removeTx(entry->GetTx().GetHash(), true)) {
+    AssertLockHeld(m_cs_fee_estimator);
```
Technically, it is no needed.

There are two reasons for that:
- if somehow annotation `EXCLUSIVE_LOCKS_REQUIRED(m_cs_fee_estimator)` will disappear (e.g., human fault) from the function declaration, the `AssertLockHeld(m_cs_fee_estimator)` will emit the `-Wthread-safety-analysis` warning
- the code below could be changed in the future, but `AssertLockHeld` in the first line will guard thread safety during both compile-time and run-time

Actually, I just followed a pattern from our [Developer Notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#threads-and-synchronization).

theStack reviewed (approved):
Code-review ACK 8c277b19c8f262e550cffe263e6d910b687ac882

Carefully reviewed the `m_cs_fee_estimator` lock acquires in `src/policy/fees.cpp` (on master first), concluded that allowing recursion was needed due to the `CBlockPolicyEstimator::removeTx(...)` method being called in `CBlockPolicyEstimator::FlushUnconfirmed(...)` and `CBlockPolicyEstimator::processBlockTx(...)` (in turn called by `CBlockPolicyEstimator::processBlock(...)`) while holding the lock already. The PR solves this by introducing a private method `_removeTx` which doesn't take the lock, and replacing the internal calls with this. Also checked that the introduced thread safety annotations make sense.

amadeuszpawlik reviewed (approved):
ACK 8c277b19c8f262e550cffe263e6d910b687ac882 reviewed, built and ran tests
